### PR TITLE
All data int test branch, mangle helper function, read me changes

### PR DESCRIPTION
**Pull request by raahi-chada2004**

### CODE REVIEW

File: integration_test.go [hunk 1]
```diff
@@ -589,3 +626,92 @@ func generateData(numRows int, sleepTime time.Duration, sendBadRow bool) error {
...
+				Start: 139830307704277524,
+				End:   139830307704277525,
+			},
+			JSONField: string(jsonBytes),
+		}
+		entry, err := json.Marshal(curr)
+		if err != nil {
+			return err
+		}
+		logger.Println(string(entry))
```
Comment (agrawal-siddharth) [inline integration_test.go:716]: Is this needed?

File: out_writeapi.go [hunk 2]
```diff
@@ -78,11 +78,46 @@ const (
 	exactlyOnceDefault         = false
 	queueRequestScalingPercent = 0.8
 	numRetriesDefault          = 4
+	dataTimeDefault            = true
```
Comment (agrawal-siddharth) [inline out_writeapi.go]: Should this be dateTimeDefault?

File: out_writeapi.go [hunk 3]
```diff
@@ -476,6 +511,13 @@ func FLBPluginInit(plugin unsafe.Pointer) int {
 		log.Printf("Invalid Max_Queue_Bytes parameter in configuration file: %s", err)
 		return output.FLB_ERROR
 	}
+
+	dataTimeStringType, err := getConfigField(plugin, "DateTime_String_Type", dataTimeDefault)
```
Comment (agrawal-siddharth) [inline out_writeapi.go]: Rename to dateTimeStringType?

**Review by agrawal-siddharth — CHANGES_REQUESTED** [review]

File: out_writeapi.go (hunk 2 above)
Comment (raahi-chada2004) [reply out_writeapi.go]: Yes, done.

File: out_writeapi.go (hunk 3 above)
Comment (raahi-chada2004) [reply out_writeapi.go]: Done.

File: integration_test.go (hunk 1 above)
Comment (raahi-chada2004) [reply integration_test.go:716]: Yes this is how we've been send the data

Comment (yifatgortler) [reply integration_test.go:716]: Can you add a comment to explain that?
It's not a common use case for logger and won't be obvious.

File: out_writeapi.go [hunk 4]
```diff
@@ -78,11 +78,46 @@ const (
...
 	numRetriesDefault          = 4
+	dateTimeDefault            = true
 )
 
+// this function mangles the top-level and complex (struct) BigQuery schema to convert NUMERIC, BIGNUMERIC, DATETIME, TIME, and JSON fields to STRING.
+func mangleInputSchema(input *storagepb.TableSchema, dataTimeString bool) *storagepb.TableSchema {
+	if input == nil {
+		return nil
+	}
+	//create a clone of the table schema
```
Comment (yifatgortler) [inline out_writeapi.go]: A comment should start with one space before the beginning of the sentence, and the first letter should be capitalized (where appropriate). i.e. `// Create`.
I assume that once we move the code to the proper repository, the built-in linter will catch all of those comments, but we shouldn't wait.
Can you fix just the comments you added in this PR?

File: integration_test.go [hunk 5]
```diff
@@ -94,7 +115,7 @@ func TestPipeline(t *testing.T) {
 
 	//Wait for fluent-bit connection to generate data; add delays before ending fluent-bit process
 	time.Sleep(2 * time.Second)
-	if err := generateData(numRows, time.Second, false); err != nil {
```
Comment (yifatgortler) [inline integration_test.go:97]: Is `generateData` being called from anywhere else? Maybe it can be deleted?

File: integration_test.go [hunk 6]
```diff
@@ -589,3 +626,92 @@ func generateData(numRows int, sleepTime time.Duration, sendBadRow bool) error {
...
+	FloatField      float64 `json:"FloatField"`
+	NumericField    string  `json:"NumericField"`
+	BigNumericField string  `json:"BigNumericField"`
+	BooleanField    bool    `json:"BooleanField"`
+	TimestampField  int64   `json:"TimestampField"`
+	DateField       int32   `json:"DateField"`
+	TimeField       string  `json:"TimeField"`
+	DateTimeField   string  `json:"DateTimeField"`
+	GeographyField  string  `json:"GeographyField"`
+	RecordField     struct {
```
Comment (yifatgortler) [inline integration_test.go:648]: Can you add another sub field to the record with a type that will verify the recursive call to `mangleInputSchema` with a changed type, such as `DateTime` or `Numeric`?

File: integration_test.go [hunk 7]
```diff
@@ -124,7 +145,23 @@ func TestPipeline(t *testing.T) {
 			t.Fatalf("Failed to read query results: %v", err)
 		}
 
+		//Verify the value of the data
```
Comment (yifatgortler) [inline integration_test.go]: Is there a way to get the number of fields in the struct `log_entry_alltypes`? If so, the first check should be that number of fields in that struct is equal to the size of `BQvalues`. That can be a good indicator of what went wrong if it fails, and that will make sure that future changes will have a proper verification e2e.

**Review by yifatgortler — CHANGES_REQUESTED** [review]

File: integration_test.go (hunk 5 above)
Comment (raahi-chada2004) [reply integration_test.go:97]: We call generateData in three tests and then call generateAllDataTypes in the fourth test. This actually sends the data to the sink (log file).

File: integration_test.go (hunk 1 above)
Comment (raahi-chada2004) [reply integration_test.go:716]: Done

File: out_writeapi.go (hunk 4 above)
Comment (raahi-chada2004) [reply out_writeapi.go]: Done.

File: integration_test.go (hunk 6 above)
Comment (raahi-chada2004) [reply integration_test.go:648]: Yes this is already done: https://github.com/raahi-chada2004/out_writeapi/blob/eb3ff4a8d2068a6329d0fff1499d8052c903d3b9/integration_test.go#L83 - the second subfield is of type numeric (and the data input is string for numeric type).

File: integration_test.go (hunk 7 above)
Comment (raahi-chada2004) [reply integration_test.go]: Done. I'm assuming this is unnecessary to add for the tests which only send one field, right?

File: integration_test.go (hunk 6 above)
Comment (yifatgortler) [reply integration_test.go:648]: Oh, I missed it. Thanks for clarifying!

File: integration_test.go (hunk 7 above)
Comment (yifatgortler) [reply integration_test.go]: Yeah, no need in other tests. Only here when you have a huge struct and it's easy for things to get missed.

**Review by yifatgortler — APPROVED** [review]

**Pull request merged**